### PR TITLE
Cut dependency between plugins and xmlserializer

### PR DESCRIPTION
**Pull request by krocard**
Add CElement::getXmlNameAttribute() method to isolate the plugins from
xmlserializer.

This patch comes in preparation to exporting the necessary symbols to plugins
in Windows environment (using export macros).

<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/01org/parameter-framework/pull/205%23issuecomment-139570639%22%2C%20%22https%3A//github.com/01org/parameter-framework/pull/205%23issuecomment-139570680%22%2C%20%22https%3A//github.com/01org/parameter-framework/pull/205%23issuecomment-139574359%22%2C%20%22https%3A//github.com/01org/parameter-framework/pull/205%23discussion_r39282583%22%2C%20%22https%3A//github.com/01org/parameter-framework/pull/205%23discussion_r39282678%22%2C%20%22https%3A//github.com/01org/parameter-framework/pull/205%23discussion_r39282695%22%2C%20%22https%3A//github.com/01org/parameter-framework/pull/205%23discussion_r39282827%22%2C%20%22https%3A//github.com/01org/parameter-framework/pull/205%23discussion_r39283112%22%2C%20%22https%3A//github.com/01org/parameter-framework/pull/205%23issuecomment-139576848%22%2C%20%22https%3A//github.com/01org/parameter-framework/pull/205%23discussion_r39283677%22%2C%20%22https%3A//github.com/01org/parameter-framework/pull/205%23discussion_r39283688%22%2C%20%22https%3A//github.com/01org/parameter-framework/pull/205%23discussion_r39283710%22%2C%20%22https%3A//github.com/01org/parameter-framework/pull/205%23discussion_r39283747%22%2C%20%22https%3A//github.com/01org/parameter-framework/pull/205%23discussion_r39283761%22%5D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/01org/parameter-framework/pull/205%23issuecomment-139570639%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22%3A%2B1%3A%20%40miguelgaio%20Please%20test%20%26%20review.%22%2C%20%22created_at%22%3A%20%222015-09-11T15%3A08%3A33Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6862950%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/krocard%22%7D%7D%2C%20%7B%22body%22%3A%20%22%40OznOg%20%40dawagner%20Please%20review%22%2C%20%22created_at%22%3A%20%222015-09-11T15%3A08%3A43Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6862950%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/krocard%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%20%22%2C%20%22created_at%22%3A%20%222015-09-11T15%3A21%3A51Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/820313%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/OznOg%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A-1%3A%20%22%2C%20%22created_at%22%3A%20%222015-09-11T15%3A32%3A12Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6430928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dawagner%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%2C%20%22Pull%20274c20ae17454198a26c47534038c112edb61ecc%20parameter/NamedElementBuilderTemplate.h%203%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/01org/parameter-framework/pull/205%23discussion_r39283112%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Update%20this%20file%27s%20copyright%20year%20too.%22%2C%20%22created_at%22%3A%20%222015-09-11T15%3A30%3A51Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6430928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dawagner%22%7D%7D%2C%20%7B%22body%22%3A%20%22Done%20%3A%2B1%3A%20%22%2C%20%22created_at%22%3A%20%222015-09-11T15%3A36%3A23Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6862950%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/krocard%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20parameter/NamedElementBuilderTemplate.h%3AL31-69%22%7D%2C%20%22Pull%20274c20ae17454198a26c47534038c112edb61ecc%20parameter/NamedElementBuilderTemplate.cpp%202%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/01org/parameter-framework/pull/205%23discussion_r39282583%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%222015%22%2C%20%22created_at%22%3A%20%222015-09-11T15%3A25%3A58Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6430928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dawagner%22%7D%7D%2C%20%7B%22body%22%3A%20%22Done%20%3A%2B1%3A%20%22%2C%20%22created_at%22%3A%20%222015-09-11T15%3A37%3A05Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6862950%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/krocard%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20parameter/NamedElementBuilderTemplate.cpp%3AL1-41%22%7D%2C%20%22Pull%20274c20ae17454198a26c47534038c112edb61ecc%20parameter/NamedElementBuilderTemplate.h%2036%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/01org/parameter-framework/pull/205%23discussion_r39282827%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22should%20we%20add%20the%20%5C%22override%5C%22%20keyword%20%3F%22%2C%20%22created_at%22%3A%20%222015-09-11T15%3A28%3A11Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6430928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dawagner%22%7D%7D%2C%20%7B%22body%22%3A%20%22Done%20%3A%2B1%3A%20%22%2C%20%22created_at%22%3A%20%222015-09-11T15%3A36%3A30Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6862950%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/krocard%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20parameter/NamedElementBuilderTemplate.h%3AL31-69%22%7D%2C%20%22Pull%20274c20ae17454198a26c47534038c112edb61ecc%20parameter/NamedElementBuilderTemplate.h%2013%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/01org/parameter-framework/pull/205%23discussion_r39282678%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22hidden%22%2C%20%22created_at%22%3A%20%222015-09-11T15%3A27%3A01Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6430928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dawagner%22%7D%7D%2C%20%7B%22body%22%3A%20%22Done%20%3A%2B1%3A%20%5Cr%5Cn%22%2C%20%22created_at%22%3A%20%222015-09-11T15%3A36%3A57Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6862950%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/krocard%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20parameter/NamedElementBuilderTemplate.h%3AL31-69%22%7D%2C%20%22Pull%20274c20ae17454198a26c47534038c112edb61ecc%20parameter/NamedElementBuilderTemplate.h%2010%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/01org/parameter-framework/pull/205%23discussion_r39282695%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22that%20-%3E%20who%22%2C%20%22created_at%22%3A%20%222015-09-11T15%3A27%3A09Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6430928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dawagner%22%7D%7D%2C%20%7B%22body%22%3A%20%22those%20that%20-%3E%20sources%22%2C%20%22created_at%22%3A%20%222015-09-11T15%3A36%3A45Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6862950%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/krocard%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20parameter/NamedElementBuilderTemplate.h%3AL31-69%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'></a>
- [ ] <a href='#crh-comment-General Comment'></a> &nbsp;<b><a href='https://github.com/01org/parameter-framework/pull/205#issuecomment-139570639'>General Comment</a></b>
- <a href='https://github.com/krocard'></a> :+1: @miguelgaio Please test & review.
- <a href='https://github.com/krocard'></a> @OznOg @dawagner Please review
- <a href='https://github.com/OznOg'></a> :+1:
- <a href='https://github.com/dawagner'></a> :-1:
- [x] <a href='#crh-comment-Pull 274c20ae17454198a26c47534038c112edb61ecc parameter/NamedElementBuilderTemplate.h 10'></a> &nbsp;<b><a href='https://github.com/01org/parameter-framework/pull/205#discussion_r39282695'>File: parameter/NamedElementBuilderTemplate.h:L31-69</a></b>
- <a href='https://github.com/dawagner'></a> that -> who
- <a href='https://github.com/krocard'></a> those that -> sources
- [x] <a href='#crh-comment-Pull 274c20ae17454198a26c47534038c112edb61ecc parameter/NamedElementBuilderTemplate.cpp 2'></a> &nbsp;<b><a href='https://github.com/01org/parameter-framework/pull/205#discussion_r39282583'>File: parameter/NamedElementBuilderTemplate.cpp:L1-41</a></b>
- <a href='https://github.com/dawagner'></a> 2015
- <a href='https://github.com/krocard'></a> Done :+1:
- [x] <a href='#crh-comment-Pull 274c20ae17454198a26c47534038c112edb61ecc parameter/NamedElementBuilderTemplate.h 13'></a> &nbsp;<b><a href='https://github.com/01org/parameter-framework/pull/205#discussion_r39282678'>File: parameter/NamedElementBuilderTemplate.h:L31-69</a></b>
- <a href='https://github.com/dawagner'></a> hidden
- <a href='https://github.com/krocard'></a> Done :+1:
- [x] <a href='#crh-comment-Pull 274c20ae17454198a26c47534038c112edb61ecc parameter/NamedElementBuilderTemplate.h 36'></a> &nbsp;<b><a href='https://github.com/01org/parameter-framework/pull/205#discussion_r39282827'>File: parameter/NamedElementBuilderTemplate.h:L31-69</a></b>
- <a href='https://github.com/dawagner'></a> should we add the "override" keyword ?
- <a href='https://github.com/krocard'></a> Done :+1:
- [x] <a href='#crh-comment-Pull 274c20ae17454198a26c47534038c112edb61ecc parameter/NamedElementBuilderTemplate.h 3'></a> &nbsp;<b><a href='https://github.com/01org/parameter-framework/pull/205#discussion_r39283112'>File: parameter/NamedElementBuilderTemplate.h:L31-69</a></b>
- <a href='https://github.com/dawagner'></a> Update this file's copyright year too.
- <a href='https://github.com/krocard'></a> Done :+1:

<a href='https://www.codereviewhub.com/01org/parameter-framework/pull/205?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/01org/parameter-framework/pull/205?approve=1'></a>&nbsp;<a href='https://github.com/01org/parameter-framework/pull/205'></a>
<a href='#crh-end'></a>

### CODE REVIEW

**Comment by krocard** [comment]
:+1: @miguelgaio Please test & review.

**Comment by krocard** [comment]
@OznOg @dawagner Please review

**Comment by OznOg** [comment]
:+1:

File: parameter/NamedElementBuilderTemplate.cpp [hunk 1]
```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright (c) 2011-2014, Intel Corporation
```
Comment (dawagner) [inline parameter/NamedElementBuilderTemplate.cpp]: 2015

File: parameter/NamedElementBuilderTemplate.h [hunk 2]
```diff
@@ -31,14 +31,38 @@
...
+namespace detail
+{
+/** Part of the implementation of `TNamedElementBuilderTemplate` that does not need
+ * to be template.
+ *
+ * If this was implemented in this header, the xmlElement implementation
+ * would leak to those that include this header.
+ *
+ * Plugins are including this header. As a result they would include
+ * the xmlElement implementation if it was not hided in a cpp.
```
Comment (dawagner) [inline parameter/NamedElementBuilderTemplate.h]: hidden

File: parameter/NamedElementBuilderTemplate.h [hunk 3]
```diff
@@ -31,14 +31,38 @@
 
 #include "ElementBuilder.h"
 
+namespace detail
+{
+/** Part of the implementation of `TNamedElementBuilderTemplate` that does not need
+ * to be template.
+ *
+ * If this was implemented in this header, the xmlElement implementation
+ * would leak to those that include this header.
```
Comment (dawagner) [inline parameter/NamedElementBuilderTemplate.h]: that -> who

File: parameter/NamedElementBuilderTemplate.h [hunk 4]
```diff
@@ -31,14 +31,38 @@
...
+std::string getName(const CXmlElement& xmlElement);
+}
+
 template <class ElementType>
 class TNamedElementBuilderTemplate : public CElementBuilder
 {
 public:
-    TNamedElementBuilderTemplate() : CElementBuilder() {}
-
     virtual CElement* createElement(const CXmlElement& xmlElement) const
```
Comment (dawagner) [inline parameter/NamedElementBuilderTemplate.h]: should we add the "override" keyword ?

File: parameter/NamedElementBuilderTemplate.h [hunk 5]
```diff
@@ -31,14 +31,38 @@
 
 #include "ElementBuilder.h"
```
Comment (dawagner) [inline parameter/NamedElementBuilderTemplate.h]: Update this file's copyright year too.

**Comment by dawagner** [comment]
:-1:

File: parameter/NamedElementBuilderTemplate.h (hunk 5 above)
Comment (krocard) [inline parameter/NamedElementBuilderTemplate.h]: Done :+1:

File: parameter/NamedElementBuilderTemplate.h (hunk 4 above)
Comment (krocard) [inline parameter/NamedElementBuilderTemplate.h]: Done :+1:

File: parameter/NamedElementBuilderTemplate.h (hunk 3 above)
Comment (krocard) [inline parameter/NamedElementBuilderTemplate.h]: those that -> sources

File: parameter/NamedElementBuilderTemplate.h (hunk 2 above)
Comment (krocard) [inline parameter/NamedElementBuilderTemplate.h]: Done :+1:

File: parameter/NamedElementBuilderTemplate.cpp (hunk 1 above)
Comment (krocard) [inline parameter/NamedElementBuilderTemplate.cpp]: Done :+1:

**Comment by krocard** [comment]
:+1:

**Pull request merged**